program.c

```diff
@@ -192,7 +278,21 @@ int program_execute(struct qdl_device *qdl, int (*apply)(struct qdl_device *qdl,
 			continue;
 		}
 
-		ret = apply(qdl, program, fd);
+		if (!program->sparse) {
+			ret = apply(qdl, program, fd);
+		} else {
+			program_sparse = program_load_sparse(program, fd);
+			if (!program_sparse) {
+				fprintf(stderr, "[PROGRAM] load sparse failed\n");
```

Comment:
I believe you printed an (more useful) error in each error path within program_load_sparse() already, so no need to tell the user once more.

Reply:
Hmm, you can generate a correct sparse image, but all chunks will be DON'T CARE or the chunk header will contain zero bytes of data.
On the one hand, the sparse image is validated, but on the other hand, there is nothing to flash and the program will have to be skipped.
Maybe you can suggest a better scenario?